README.md
# assertions

Assertions helps in performing assertions in testing projects also on api responses

- [Prerequisites](#Prerequisites)
- [Installation](#Installation)
- [Build library](#Build-library)
- [Launch tests](#Launch-tests)
- [Usage](#Usage)
    - [DeclarativeAssertion](#DeclarativeAssertion)
    - [ApiAssertion](#ApiAssertion)

## Prerequisites

- Java 8+
- Gradle 8+ (gradle wrapper included)

## Installation

Using Gradle

```
dependencies {
    implementation 'io.github.danieleperuzzi:assertions:1.0.0'
}
```

Using Maven

```xml
<dependency>
  <groupId>io.github.danieleperuzzi</groupId>
  <artifactId>assertions</artifactId>
  <version>1.0.0</version>
</dependency>
```

## Build library

To build assertions library just run gradle build task:

on Linux
```
./gradlew build
```

on Windows
```
./gradlew.bat build
```

you can also build the library using your machine gradle installation but please be sure gradle version is at least 8.

## Launch tests

To launch the builtin test suite for assertions library just run gradle test task:

on Linux
```
./gradlew test
```

on Windows
```
./gradlew.bat test
```

you can also test the library using your machine gradle installation but please be sure gradle version is at least 8.

## Usage

### DeclarativeAssertion

A more fashionable way to perform assertions if the condition is met or not

```java
import static io.github.danieleperuzzi.assertion.DeclarativeAssertion.test; // used for readability

ApiResponse apiResponse;

test(apiResponse)
        .when(response -> response.getStatus() == 200)
        .then(response -> assertEquals("OK", response.getResponseBody().getStatus()));

test(apiResponse)
        .when(response -> response.getStatus() == 400)
        .then(response -> assertEquals("Error", response.getResponseBody().getStatus()));
```

suppose api status is ```200``` then only the first assertion is performed

### ApiAssertion

Since every API has error handling it responds in different ways depending it is successful or failure and the structure 
of the response may change depending on that.

```ApiAssertion``` helps in performing right checks when API is succesful or not

```java
ApiResponse apiResponse;

try {
    new ApiAssertion<>(apiResponse)
        .isSuccessful(response -> response.getStatus() == 200)
        .onSuccess(response -> assertEquals("OK", response.getResponseBody().getStatus()))
        .onFailure(response -> assertEquals("KO", response.getResponseBody().getStatus()))
        .test();
} catch (Exception e) {
    e.printStackTrace();
}
```

depending on the api status then rights checks are performed

In case we are dealing with multiple error cases we can use conditional ```onFailure``` to elegantly handle them

```java
try {
    new ApiAssertion<>(apiResponse)
        .isSuccessful(response -> response.getStatus() == 200)
        .onSuccess(response -> assertEquals("OK", response.getResponseBody().getStatus()))
        .onFailure(response -> response.getStatus() == 400, response -> assertEquals("Error 400", response.getResponseBody().getStatus()))
        .onFailure(response -> response.getStatus() == 401, response -> assertEquals("Error 401", response.getResponseBody().getStatus()))
        .test();
} catch (Exception e) {
    e.printStackTrace();
}
```

whenever api status is failure then the right check is performed depending on its status code


CHANGELOG.md
# CHANGELOG

## 1.0.0

#### Added:

- ```DeclarativeAssertion``` class to perform assertions on objects in a declarative way
- ```ApiAssertion``` class to perform assertions specifically on API responses

#### Added tests:

- ```DeclarativeAssertion```
- ```ApiAssertion```

lib/src/test/java/com/danieleperuzzi/assertion/api/ApiAssertionTest.java
/*
 * Copyright 2023 Daniele Peruzzi. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.danieleperuzzi.assertion.api;

import com.danieleperuzzi.assertion.util.ApiResponseMock;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

@ExtendWith(MockitoExtension.class)
public class ApiAssertionTest {

    private static ApiResponseMock apiResponseOk;
    private static ApiResponseMock apiResponseKo;

    @BeforeAll
    public static void staticSetUp() {
        apiResponseOk = new ApiResponseMock(200, "{\"status\": \"OK\", \"message\": \"response is successful\"}");
        apiResponseKo = new ApiResponseMock(400, "{\"status\": \"KO\", \"message\": \"response is failure\"}");
    }

    @Test
    @DisplayName("define API predicate exception")
    public void defineApiPredicateException() {
        Exception exception = assertThrows(Exception.class, () -> {
            new ApiAssertion<>(apiResponseOk)
                    .test();
        });

        String expectedMessage = "Define at least API predicate";
        String actualMessage = exception.getMessage();

        assertEquals(expectedMessage, actualMessage);
    }

    @Test
    @DisplayName("define API onSuccess or onFailure exception")
    public void defineApiOnSuccessOnFailureException() {
        Exception exception = assertThrows(Exception.class, () -> {
            new ApiAssertion<>(apiResponseOk)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .test();
        });

        String expectedMessage = "Define at least API onSuccess or onFailure assertions";
        String actualMessage = exception.getMessage();

        assertEquals(expectedMessage, actualMessage);
    }

    @Test
    @DisplayName("define only one API isSuccessful predicate exception")
    public void defineOnlyOneApiIsSuccessfulException() {
        Exception exception = assertThrows(Exception.class, () -> {
            new ApiAssertion<>(apiResponseOk)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .isSuccessful(r -> r.getStatus() == 204)
                    .test();
        });

        String expectedMessage = "Define only one isSuccessful predicate";
        String actualMessage = exception.getMessage();

        assertEquals(expectedMessage, actualMessage);
    }

    @Test
    @DisplayName("define only one API onSuccess assertion exception")
    public void defineOnlyOneApiOnSuccessException() {
        AtomicInteger testOk = new AtomicInteger(0);

        Exception exception = assertThrows(Exception.class, () -> {
            new ApiAssertion<>(apiResponseOk)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onSuccess(r -> testOk.incrementAndGet())
                    .onSuccess(r -> testOk.incrementAndGet())
                    .test();
        });

        String expectedMessage = "Define only one onSuccess assertion";
        String actualMessage = exception.getMessage();

        assertEquals(expectedMessage, actualMessage);
    }

    @Test
    @DisplayName("define only one API onFailure assertion exception")
    public void defineOnlyOneApiOnFailureException() {
        AtomicInteger testKo = new AtomicInteger(0);

        Exception exception = assertThrows(Exception.class, () -> {
            new ApiAssertion<>(apiResponseOk)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onFailure(r -> testKo.incrementAndGet())
                    .onFailure(r -> testKo.incrementAndGet())
                    .test();
        });

        String expectedMessage = "Define only one onFailure assertion";
        String actualMessage = exception.getMessage();

        assertEquals(expectedMessage, actualMessage);
    }

    @Test
    @DisplayName("define only one API onFailure assertion or conditional onFailure assertions exception")
    public void defineOnlyOneApiOnFailureOrConditionalOnFailureException() {
        AtomicInteger testKo = new AtomicInteger(0);

        Exception exception = assertThrows(Exception.class, () -> {
            new ApiAssertion<>(apiResponseOk)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onFailure(r -> testKo.incrementAndGet())
                    .onFailure(r -> r.getStatus() == 400, r -> testKo.incrementAndGet())
                    .test();
        });

        String expectedMessage = "Define only simple or conditional failure assertions";
        String actualMessage = exception.getMessage();

        assertEquals(expectedMessage, actualMessage);
    }

    @Test
    @DisplayName("check onSuccess")
    public void checkOnSuccess() {
        AtomicInteger testOk = new AtomicInteger(0);

        try {
            new ApiAssertion<>(apiResponseOk)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onSuccess(r -> testOk.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(1, testOk.get());

        try {
            new ApiAssertion<>(apiResponseKo)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onSuccess(r -> testOk.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(1, testOk.get());
    }

    @Test
    @DisplayName("check onFailure")
    public void checkOnFailure() {
        AtomicInteger testKo = new AtomicInteger(0);

        try {
            new ApiAssertion<>(apiResponseOk)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onFailure(r -> testKo.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(0, testKo.get());

        try {
            new ApiAssertion<>(apiResponseKo)
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onFailure(r -> testKo.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(1, testKo.get());
    }

    @Test
    @DisplayName("check conditional onFailure")
    public void checkConditionalOnFailure() {
        AtomicInteger testKo400 = new AtomicInteger(0);
        AtomicInteger testKo401 = new AtomicInteger(0);

        try {
            new ApiAssertion<>(apiResponseOk) // check response KO
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onFailure(r -> r.getStatus() == 400, r -> testKo400.incrementAndGet())
                    .onFailure(r -> r.getStatus() == 401, r -> testKo401.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(0, testKo400.get());
        assertEquals(0, testKo401.get());

        try {
            new ApiAssertion<>(apiResponseKo) // check response KO
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onFailure(r -> r.getStatus() == 400, r -> testKo400.incrementAndGet())
                    .onFailure(r -> r.getStatus() == 401, r -> testKo401.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(1, testKo400.get());
        assertEquals(0, testKo401.get());
    }

    @Test
    @DisplayName("check onSuccess and onFailure")
    public void checkOnSuccessOnFailure() {
        AtomicInteger testOk = new AtomicInteger(0);
        AtomicInteger testKo = new AtomicInteger(0);

        try {
            new ApiAssertion<>(apiResponseOk) // check response OK
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onSuccess(r -> testOk.incrementAndGet())
                    .onFailure(r -> testKo.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(1, testOk.get());
        assertEquals(0, testKo.get());


        try {
            new ApiAssertion<>(apiResponseKo) // check response KO
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onSuccess(r -> testOk.incrementAndGet())
                    .onFailure(r -> testKo.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(1, testOk.get());
        assertEquals(1, testKo.get());
    }

    @Test
    @DisplayName("check onSuccess and conditional onFailure")
    public void checkOnSuccessConditionalOnFailure() {
        AtomicInteger testOk = new AtomicInteger(0);
        AtomicInteger testKo400 = new AtomicInteger(0);
        AtomicInteger testKo401 = new AtomicInteger(0);

        try {
            new ApiAssertion<>(apiResponseOk) // check response KO
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onSuccess(r -> testOk.incrementAndGet())
                    .onFailure(r -> r.getStatus() == 400, r -> testKo400.incrementAndGet())
                    .onFailure(r -> r.getStatus() == 401, r -> testKo401.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(1, testOk.get());
        assertEquals(0, testKo400.get());
        assertEquals(0, testKo401.get());

        try {
            new ApiAssertion<>(apiResponseKo) // check response KO
                    .isSuccessful(r -> r.getStatus() == 200)
                    .onSuccess(r -> testOk.incrementAndGet())
                    .onFailure(r -> r.getStatus() == 400, r -> testKo400.incrementAndGet())
                    .onFailure(r -> r.getStatus() == 401, r -> testKo401.incrementAndGet())
                    .test();
        } catch (Exception e) {
            e.printStackTrace();
        }

        assertEquals(1, testOk.get());
        assertEquals(1, testKo400.get());
        assertEquals(0, testKo401.get());
    }
}


lib/src/test/java/com/danieleperuzzi/assertion/util/ApiResponseMock.java
/*
 * Copyright 2023 Daniele Peruzzi. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.danieleperuzzi.assertion.util;

public class ApiResponseMock {

    private int status;
    private String responseText;

    public ApiResponseMock(int status, String responseText) {
        this.status = status;
        this.responseText = responseText;
    }

    public int getStatus() {
        return status;
    }

    public String getResponseText() {
        return responseText;
    }
}


lib/src/test/java/com/danieleperuzzi/assertion/DeclarativeAssertionTest.java
/*
 * Copyright 2023 Daniele Peruzzi. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.danieleperuzzi.assertion;

import com.danieleperuzzi.assertion.util.ApiResponseMock;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.concurrent.atomic.AtomicInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static com.danieleperuzzi.assertion.DeclarativeAssertion.test;

@ExtendWith(MockitoExtension.class)
public class DeclarativeAssertionTest {

    private static ApiResponseMock apiResponseOk;

    @BeforeAll
    public static void staticSetUp() {
        apiResponseOk = new ApiResponseMock(200, "{\"status\": \"OK\", \"message\": \"response is successful\"}");
    }

    @Test
    @DisplayName("predicate is success")
    public void successPredicate() {
        AtomicInteger testOk = new AtomicInteger(0);

        test(apiResponseOk)
                .when(response -> response.getStatus() == 200)
                .then(response -> testOk.incrementAndGet());

        assertEquals(1, testOk.get());
    }

    @Test
    @DisplayName("predicate is failure")
    public void failurePredicate() {
        AtomicInteger testKo = new AtomicInteger(0);

        test(apiResponseOk)
                .when(response -> response.getStatus() == 400)
                .then(response -> testKo.incrementAndGet());

        assertEquals(0, testKo.get());
    }
}


lib/src/main/java/io/github/danieleperuzzi/assertion/DeclarativeAssertion.java
/*
 * Copyright 2023 Daniele Peruzzi. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.danieleperuzzi.assertion;

import java.util.function.Consumer;
import java.util.function.Predicate;

/**
 * Simple class to perform some conditional test in declarative and easy readable mode.
 *
 * The result of the application of this class would be
 *
 * <pre>{@code
 * test(apiResponse)
 *      .when(response -> response.getStatus() == 200)
 *      .then(response -> validateSuccessResponse(response));
 * }</pre>
 *
 * @param <T>   the type of the object to be tested
 */
public class DeclarativeAssertion<T> {

    private T object;
    private Predicate<T> predicate;

    private DeclarativeAssertion(T object) {
        this.object = object;
    }

    /**
     * Used to create a new DeclarativeAssertion instance, the main purpose is to wrap object
     * instantiation with a method with an explicit name
     *
     * @param object    the object to be tested
     * @return          a new DeclarativeAssertion instance
     * @param <T>       the type of the object to be tested
     */
    public static <T> DeclarativeAssertion<T> test(T object) {
        return new DeclarativeAssertion<>(object);
    }

    /**
     * Store this predicate for later use
     *
     * @param p     the predicate that tests the object
     * @return      current DeclarativeAssertion instance
     */
    public DeclarativeAssertion<T> when(Predicate<T> p) {
        predicate = p;

        return this;
    }

    /**
     * Perform the test on the stored object only if the given
     * predicate is successful
     *
     * @param c     a consumer that represents the assertions on
     *              the stored object
     */
    public void then(Consumer<T> c) {
        if (predicate.test(object)) {
            c.accept(object);
        }
    }
}


lib/src/main/java/io/github/danieleperuzzi/assertion/api/ApiAssertion.java
/*
 * Copyright 2023 Daniele Peruzzi. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.danieleperuzzi.assertion.api;

import java.util.*;
import java.util.function.Predicate;
import java.util.function.Consumer;

/**
 * This class helps in performing assertions specifically on response APIs because the pattern is always the same: on check
 * for a success case and multiple checks depending on the error triggered
 *
 * The result of the application of this api assertion would be
 *
 * <pre>{@code
 * new ApiAssertion<>(apiResponse)
 *      .isSuccessful(r -> r.getStatus() == 200)
 *      .onSuccess(r -> testOk(r))
 *      .onFailure(r -> r.getStatus() == 400, r -> testKo400(r))
 *      .onFailure(r -> r.getStatus() == 401, r -> testKo401(r))
 *      .test();
 * }</pre>
 *
 * @param <R>   the type of the API response to be tested
 */
public class ApiAssertion<R> {

    private R response;
    private Predicate<R> isSuccessfulPredicate;
    private Consumer<R> successAssertions;
    private Consumer<R> failureAssertions;
    private Map<Predicate<R>, Consumer<R>> failureAssertionMap = new HashMap<>();

    /**
     * Creates a new ApiAssertion
     *
     * @param response      the API response to test
     */
    public ApiAssertion(R response) {
        this.response = response;
        this.isSuccessfulPredicate = null;
        this.successAssertions = null;
        this.failureAssertions = null;
    }

    /**
     * Defines the predicate that checks whenever the API response to be tested is a successful response
     * or a failure response
     *
     * @param p             the predicate that performs the is successful test on the API response
     * @return              this class instance to chain more actions
     * @throws Exception    exception thrown in case we already defined one isSuccessful predicate
     */
    public ApiAssertion<R> isSuccessful(Predicate<R> p) throws Exception {
        if (!Objects.isNull(isSuccessfulPredicate)) {
            throw new Exception("Define only one isSuccessful predicate");
        }

        isSuccessfulPredicate = p;

        return this;
    }

    /**
     * Defines the assertions to be performed when the API response to be tested is a successful response
     *
     * @param ok            the test to be performed in case the API response is success
     * @return              this class instance to chain more actions
     * @throws Exception    exception thrown in case we already defined one onSuccess assertion
     */
    public ApiAssertion<R> onSuccess(Consumer<R> ok) throws Exception {
        if (!Objects.isNull(successAssertions)) {
            throw new Exception("Define only one onSuccess assertion");
        }

        successAssertions = ok;

        return this;
    }

    /**
     * Defines the assertions to be performed when the API response to be tested is a failure response
     *
     * @param ko            the test to be performed in case the API response is failure
     * @return              this class instance to chain more actions
     * @throws Exception    exception thrown in case we already defined one onFailure assertion
     */
    public ApiAssertion<R> onFailure(Consumer<R> ko) throws Exception {
        if (!Objects.isNull(failureAssertions)) {
            throw new Exception("Define only one onFailure assertion");
        }

        failureAssertions = ko;

        return this;
    }

    /**
     * Defines multiple one failure assertions based on specific conditions. This is useful because the api response can fail
     * in multiple ways so assertions may change between error cases
     *
     * @param p     the predicate that checks if the current assertions are the right ones for this specific error case
     * @param ko    the test to be performed in case the API response is failure and the predicate is satisfied
     * @return      this class instance to chain more actions
     */
    public ApiAssertion<R> onFailure(Predicate<R> p, Consumer<R> ko) {
        failureAssertionMap.put(p, ko);

        return this;
    }

    private void testOk(R response) {
        Optional.of(response)
                .filter(r -> isSuccessfulPredicate.test(r))
                .ifPresent(r -> successAssertions.accept(r));
    }

    private void testConditionalKo(R response) {
        failureAssertionMap.entrySet()
                .stream()
                .filter(entry -> entry.getKey().test(response))
                .forEach(entry -> entry.getValue().accept(response));
    }

    private void testKo(R response) {
        Optional.of(response)
                .filter(r -> !isSuccessfulPredicate.test(r) && !Objects.isNull(failureAssertions) && failureAssertionMap.size() == 0)
                .ifPresent(r -> failureAssertions.accept(r));

        Optional.of(response)
                .filter(r -> !isSuccessfulPredicate.test(r) && Objects.isNull(failureAssertions) && failureAssertionMap.size() > 0)
                .ifPresent(this::testConditionalKo);
    }

    private void testOkKo(R response) {
        testOk(response);
        testKo(response);
    }

    /**
     * Perform the API response test
     *
     * @throws Exception    exceptions are thrown in these cases:
     *
     *                      <ul>
     *                          <li>the is successful check isn't defined</li>
     *                          <li>no success or failure assertions are defined</li>
     *                          <li>both simple and conditional on failure assertions are defined</li>
     *                      </ul>
     */
    public void test() throws Exception {
        Optional.ofNullable(isSuccessfulPredicate)
                .orElseThrow(() -> new Exception("Define at least API predicate"));

        if (Objects.isNull(successAssertions) && Objects.isNull(failureAssertions) && failureAssertionMap.size() == 0) {
            throw new Exception("Define at least API onSuccess or onFailure assertions");
        }

        if (!Objects.isNull(failureAssertions) && failureAssertionMap.size() > 0) {
            throw new Exception("Define only simple or conditional failure assertions");
        }

        // perform success and failure assertions
        if (!Objects.isNull(successAssertions) && (!Objects.isNull(failureAssertions) || failureAssertionMap.size() > 0)) {
            testOkKo(response);
        }

        // perform only success assertions
        if (!Objects.isNull(successAssertions) && Objects.isNull(failureAssertions) && failureAssertionMap.size() == 0) {
            testOk(response);
        }

        // perform only failure assertions
        if (Objects.isNull(successAssertions) && (!Objects.isNull(failureAssertions) || failureAssertionMap.size() > 0)) {
            testKo(response);
        }
    }
}
